internal/cmdrender/executor.go

```diff
@@ -429,31 +421,70 @@ func cloneResources(input []*yaml.RNode) (output []*yaml.RNode) {
 }
 
 // path (location) of a KRM resources is tracked in a special key in
-// metadata.annotation field. adjustRelPath updates that path annotation by prepending
-// the given relPath to the current path annotation if it doesn't exist already.
-// Resources are read from local filesystem or generated at a package level, so the
-// path annotation in each resource points to path relative to that package.
-// But the resources are written to the file system at the root package level, so
+// metadata.annotation field that is used to write the resources to the filesystem.
+// When resources are read from local filesystem or generated at a package level, the
+// path annotation in a resource points to path relative to that package. But the resources
+// are written to the file system at the root package level, so
 // the path annotation in each resources needs to be adjusted to be relative to the rootPkg.
-func adjustRelPath(resources []*yaml.RNode, relPath string) ([]*yaml.RNode, error) {
-	if relPath == "" {
-		return resources, nil
-	}
+// adjustRelPath updates the path annotation by prepending the path of the package
+// a resource relative to the root package.
+func adjustRelPath(hctx *hydrationContext) error {
+	resources := hctx.root.resources
 	for _, r := range resources {
+		pkgPath, err := pkg.GetPkgPathAnnotation(r)
+		if err != nil {
+			return err
+		}
+		// Note: kioutil.GetFileAnnotation returns OS specific
+		// paths today, https://github.com/kubernetes-sigs/kustomize/issues/3749
 		currPath, _, err := kioutil.GetFileAnnotations(r)
 		if err != nil {
-			return resources, err
+			return err
+		}
+		newPath, err := pathRelToRoot(string(hctx.root.pkg.UniquePath), pkgPath, currPath)
+		if err != nil {
+			return err
 		}
-		// if currPath is relative to root pkg i.e. already has relPath, skip it
-		if !strings.HasPrefix(currPath, relPath+"/") {
-			newPath := path.Join(relPath, currPath)
-			err = r.PipeE(yaml.SetAnnotation(kioutil.PathAnnotation, newPath))
-			if err != nil {
-				return resources, err
-			}
+		if err = r.PipeE(yaml.SetAnnotation(kioutil.PathAnnotation, newPath)); err != nil {
+			return err
+		}
+		if err = pkg.RemovePkgPathAnnotation(r); err != nil {
+			return err
 		}
 	}
-	return resources, nil
+	return nil
+}
+
+// pathRelToRoot computes resource's path relative to root package given
+// absolute rootPkgPath, absolute subpkgPath and resource path relative to sub package.
+// All the inputs paths are assumed to be OS specific.
+func pathRelToRoot(rootPkgPath, subPkgPath, resourcePath string) (relativePath string, err error) {
```

Comment:
does it make sense to take in UniquePath instead of string?

Reply:
I tried that initially, it led to mix results because `subPkgPath` is originally available as string in the caller and then inside the function, we need the string form of `rootPkgPath` for other fs function calls. And this being an internal function, wasn't convinced with the final result, so reverted it.